syft/frameworks/torch/mpc/fss.py

```diff
@@ -195,210 +226,365 @@ def le(x1, x2):
+            CWi = uncompress(_CW[i], op=COMP)
+            dual_state = H(s[i]) ^ (t[i] * CWi)
+            state = multi_dim_filter(dual_state, x[i])
+            σ[i + 1], τ[i + 1], s[i + 1], t[i + 1] = split(state, (COMP, λs, 1, λs, 1))
+            out[i] = (-1) ** b * (τ[i + 1] * CW_leaf[i] + convert(σ[i + 1]))
+
+        # Last node, the other σ is also a leaf
+        out[n] = (-1) ** b * (t[n].squeeze() * CW_leaf[n] + convert(s[n]))
+
+        return out.sum(axis=0).astype(np.int64).reshape(original_shape)
+
+
+def compress(CWi, alpha_i, op=EQ):
+    if op == EQ:
+        sL, tL, sR, tR = split(CWi, (op, λs, 1, λs, 1))
+        return (tL.astype(np.bool), tR.astype(np.bool), (1 - alpha_i) * sR + alpha_i * sL)
+    else:
+        σL, τL, sL, tL, σR, τR, sR, tR = split(CWi, (op, λs, 1, λs, 1, λs, 1, λs, 1))
+        return (
+            τL.astype(np.bool),
+            tL.astype(np.bool),
+            τR.astype(np.bool),
+            tR.astype(np.bool),
+            alpha_i * σR + (1 - alpha_i) * σL,
+            (1 - alpha_i) * sR + alpha_i * sL,
+        )
+
+
+def uncompress(_CWi, op=EQ):
+    if op == EQ:
+        CWi = concat(
+            _CWi[2],
+            _CWi[0].reshape(1, -1).astype(np.uint64),
+            _CWi[2],
+            _CWi[1].reshape(1, -1).astype(np.uint64),
+        ).reshape(2, 3, -1)
+    else:
+        CWi = concat(
+            _CWi[4],
+            _CWi[0].reshape(1, -1).astype(np.uint64),
+            _CWi[5],
+            _CWi[1].reshape(1, -1).astype(np.uint64),
+            _CWi[4],
+            _CWi[2].reshape(1, -1).astype(np.uint64),
+            _CWi[5],
+            _CWi[3].reshape(1, -1).astype(np.uint64),
+        ).reshape(2, 6, -1)
+    return CWi
+
+
+def Array(*shape):
+    return np.empty(shape, dtype=np.uint64)
+
+
+def bit_decomposition(x):
+    x = x.astype(np.uint32)
+    n_values = x.shape[0]
+    x = x.reshape(-1, 1).view(np.uint8)
+    x = x.reshape(n_values, 4, 1)
+    x = x >> np.arange(8, dtype=np.uint8)
+    x = x & 0b1
+    x = np.flip(x.reshape(n_values, -1)[:, :n], axis=1).T
+    return x
+
+
+def randbit(shape):
+    assert len(shape) == 3
+    byte_dim = shape[-2]
+    shape_with_bytes = shape[:-2] + (math.ceil(byte_dim / 64), shape[-1])
+    randvalues = np.random.randint(0, 2 ** 64, size=shape_with_bytes, dtype=np.uint64)
+    randvalues[:, 0] = randvalues[:, 0] % 2 ** (byte_dim % 64)
+    return randvalues
+
+
+def concat(*args, **kwargs):
+    return np.concatenate(args, **kwargs)
+
+
+def split_last_bit(buffer):
+    # Numbers are on 64 bits
+    return buffer & 0b1111111111111111111111111111111111111111111111111111111111111110, buffer & 0b1
 
-        # Last tour, the other σ is also a leaf:
-        FnOutput[n] = t[n]
-        flat_result = FnOutput.sum(axis=0) % 2
-        return flat_result.reshape(original_shape)
+
+def huge_loop(seed_t_bytes, n_iter):
+    return [hashlib.sha3_256(seed_t_bytes[i * 16 : (i + 1) * 16]).digest() for i in range(n_iter)]
```

Comment:
Wasn't shaloop purpose to reduce this computation?

Reply:
Correct! This one is not used anymore... removing it.